README.rst
===============================================================================
Besra - image classification for protein crystallization experiments
===============================================================================

-------------------------------------------------------------------------------
About
-------------------------------------------------------------------------------

Besra is a tool for auto-classifying protein crystallization experiments. Source
and binary releases are available on `GitHub <https://github.com/ubccr/besra/releases>`_.

The goal is to implement a fast and accurate binary classifier for determining
crystal-positive vs crystal-negative images in high-throughput protein
crystallization experiments. Accuracy equal or better to that of a human would
be considered a success. Current methods take upwards of ~10 hours to classify
1536 images. Speeding up the classification will allow better integration into
existing expert knowledge systems/pipelines and enable robust evaluation and
tuning of classification algorithms across millions of images. Extending from
binary classification to n-way classification (clear, precipitate, skin, phase
separation) is supported however performance has not been extensivley tested.

-------------------------------------------------------------------------------
Quickstart
-------------------------------------------------------------------------------

First need to train on a set of images. besra-trainer requires a <TAB> separated
input file of image paths and class labels. For example, an input file with
crystal-positive = 1 and crystal-negative = 0 looks like this::

  /images/png/X0000051270868200506241635.png    1
  /images/png/X0000049750501200505270943.png    1
  /images/png/X0000049151305200506061345.png    1
  /images/png/X0000050511419200507041505.png    0
  /images/png/X0000051830553200507012227.png    0
  /images/png/X0000050611108200507051159.png    0
  ...

The path should be the full path to the image on the filesystem and the class
label should be a float. The input file requires at least 2 distinct class
labels. To train a set of images run::

  $ besra-trainer -i input.tsv -v

To speed up processing set --threads option equal to the number of cores
available on your machine. For the full set of options see::

  $ besra-trainer --help

This command will output 2 files: stats-model.xml and bow-vocab.yml which can
later be used to classify images (without having to re-train each time).

To classify a directory of images::

  $ besra-classify -i /path/to/images -m stats-model.xml -b bow-vocab.yml -v

To classify images using an input file (must be one image path per line,
similar to the input for besra-trainer)::

  $ besra-classify -i input.tsv -m stats-model.xml -b bow-vocab.yml -v

Results are written to a file named: besra-results.tsv

-------------------------------------------------------------------------------
Implementation
-------------------------------------------------------------------------------

Besra currently uses the bag-of-visual-words method [1] and a support vector
machine (SVM) classifier. Keypoints/local features are computed from the
training set using SURF [2] descriptors and clustered using k-means into a
visual vocabulary. An SVM with a linear kernel is used for image
classification.

The assumption is that the clustered features computed from crystal-positive
images will be distinct enough from crystal-negative images to produce an
accurate classifier. 

TODO:

- Optimize the parameters of SURF (hessian threshold, gaussian pyramid
  octaves, etc.). What are the appropriate settings for our data?

- Optimize the number of k-means clusters when computing the BOW
  vocabulary. Is there an optimal number of clusters?

- Experiment with other descriptor/keypoint extractors/detectors available in
  OpenCV (FAST, MSER, ORB, BRISK, etc.). See `features2d <http://docs.opencv.org/modules/features2d/doc/features2d.html>`_ 
  for the complete list.

- Experiment with different SVM types and kernels. See `svm <http://docs.opencv.org/modules/ml/doc/support_vector_machines.html>`_

- Test performance on other classes of images (clear, precipitate, phase separation). 

- Test OpenMP threads vs SURF_GPU

-------------------------------------------------------------------------------
Requirements
-------------------------------------------------------------------------------

- `boost <http://www.boost.org/>`_ >= 1.57
- `OpenCV <http://opencv.org/>`_ >= 2.4.9
- `CMake <http://www.cmake.org/>`_ >= 1.8

-------------------------------------------------------------------------------
Installation
-------------------------------------------------------------------------------

Besra uses cmake. To compile run::

  $ git clone https://github.com/ubccr/besra.git besra
  $ cd besra
  $ mkdir build
  $ cd build
  $ cmake ..
  $ make

If boost is compiled in a non-standard location run::

  $ BOOST_ROOT=/path/to/boost cmake -DBoost_NO_SYSTEM_PATHS=TRUE ..

If OpenCV is compiled in a non-standard location run::

  $ OpenCV_DIR=/path/to/opencv cmake ..

To compile besra with GPU support (requires GPU/CUDA support to be compiled in
OpenCV)::

  $ cmake -DUSE_GPU=on ..

-------------------------------------------------------------------------------
Compiling OpenCV
-------------------------------------------------------------------------------

To enable multi-threaded clustering, compile OpenCV with OpenMP support. For
example::

  $ unzip opencv-2.4.x.zip
  $ cd opencv-2.4.x/
  $ mkdir build
  $ cd build
  $ cmake -D CMAKE_BUILD_TYPE=RELEASE \
    -D CMAKE_INSTALL_PREFIX=/path/to/localdir -DWITH_OPENMP=Yes ..

For more info see `OpenCV docs <http://docs.opencv.org/trunk/doc/tutorials/introduction/linux_install/linux_install.html>`_.

-------------------------------------------------------------------------------
Compiling Boost
-------------------------------------------------------------------------------

To compile boost::

  $ tar xvf boost-1.xx.x.tar.gz
  $ cd boost_1_xx_x
  $ ./bootstrap.sh --prefix=/path/to/localdir \
       --with-libraries=log,thread,date_time,filesystem,system,program_options
  $ ./b2 install

For more info see `boost docs <http://www.boost.org/doc/libs/1_57_0/more/getting_started/unix-variants.html>`_.

-------------------------------------------------------------------------------
License
-------------------------------------------------------------------------------

Copyright (C) 2014 Andrew E. Bruno

Besra is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program.  If not, see <http://www.gnu.org/licenses/>.

-------------------------------------------------------------------------------
References
-------------------------------------------------------------------------------

[1] Csurka, Gabriella, et al. "Visual categorization with bags of keypoints."
    Workshop on statistical learning in computer vision, ECCV. Vol. 1. No. 1-22.
    2004.

[2] Bay, H. and Tuytelaars, T. and Van Gool, L. "SURF: Speeded Up Robust
    Features", 9th European Conference on Computer Vision, 2006


ChangeLog.rst
ChangeLog
=========

v0.0.3
----------------------

- Add check for supported detector/extractor combinations

v0.0.2
----------------------

- Updated to OpenCV 3.0-dev
- Add support for selecting feature detector/extractor.
- Add support for selecting classifier (SVM or Normal Bayes).

v0.0.1
----------------------

- Initial release


src/besra.cpp
/**
 * Copyright (C) 2014 Andrew E. Bruno
 *
 * This file is part of Besra
 *
 * Besra is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "besra.hpp"

namespace besra {

    void init_log(bool verbose) {
        boost::log::register_simple_formatter_factory< boost::log::trivial::severity_level, char >("Severity");
        boost::log::add_console_log(
            std::cerr, 
            boost::log::keywords::format = "%TimeStamp% [%Severity%]: %Message%",
            boost::log::keywords::auto_flush = true
        );

        boost::log::add_common_attributes();

        if(verbose) {
            boost::log::core::get()->set_filter(
                boost::log::trivial::severity >= boost::log::trivial::info
            );
        } else {
            boost::log::core::get()->set_filter(
                boost::log::trivial::severity >= boost::log::trivial::error
            );
        }

    }

    Besra::Besra(int minHessian, std::string extractor, std::string detector, bool bayes) {
        this->bayes = bayes;

        if(extractor == "SURF") {
            this->extractor = cv::xfeatures2d::SURF::create(minHessian);
        } else if(extractor == "FREAK") {
            this->extractor = cv::xfeatures2d::FREAK::create();
        } else if(extractor == "BRISK") {
            this->extractor = cv::BRISK::create();
        } else if(extractor == "ORB") {
            this->extractor = cv::ORB::create();
        } else if(extractor == "KAZE") {
            this->extractor = cv::KAZE::create();
        } else if(extractor == "AKAZE") {
            this->extractor = cv::AKAZE::create();
        } else if(extractor == "BRIEF") {
            this->extractor = cv::xfeatures2d::BriefDescriptorExtractor::create();
        } else if(extractor == "SIFT") {
            this->extractor = cv::xfeatures2d::SIFT::create();
        }

        if(detector == "SURF") {
            this->detector = cv::xfeatures2d::SURF::create(minHessian);
        } else if(detector == "BRISK") {
            this->detector = cv::BRISK::create();
        } else if(detector == "ORB") {
            this->detector = cv::ORB::create();
        } else if(detector == "KAZE") {
            this->detector = cv::KAZE::create();
        } else if(detector == "AKAZE") {
            this->detector = cv::AKAZE::create();
        } else if(detector == "MSER") {
            this->detector = cv::MSER::create();
        } else if(detector == "SIFT") {
            this->detector = cv::xfeatures2d::SIFT::create();
        } else if(detector == "FAST") {
            this->detector = cv::FastFeatureDetector::create();
        } else if(detector == "GFTT") {
            this->detector = cv::GFTTDetector::create();
        } else if(detector == "BLOB") {
            this->detector = cv::SimpleBlobDetector::create();
        }

        if(this->detector != NULL) {
            BOOST_LOG_TRIVIAL(info) << "Detector has descriptor type: " << this->detector->descriptorType();
            if(this->detector->descriptorType() == CV_32F) {
                BOOST_LOG_TRIVIAL(info) << "Float descriptor. Using BruteForce matcher";
                this->matcher = cv::DescriptorMatcher::create("BruteForce");
                //this->matcher = cv::DescriptorMatcher::create("FlannBased");
            } else {
                // assume binary?
                BOOST_LOG_TRIVIAL(info) << "Binary descriptor. Using BruteForce-Hamming matcher";
                this->matcher = cv::DescriptorMatcher::create("BruteForce-Hamming");
                /*
                cv::Ptr<cv::flann::IndexParams> indexParams = cv::makePtr<cv::flann::AutotunedIndexParams>();
                cv::Ptr<cv::flann::SearchParams> searchParams = cv::makePtr<cv::flann::SearchParams>();
                indexParams->setAlgorithm(cvflann::FLANN_INDEX_LSH);
                searchParams->setAlgorithm(cvflann::FLANN_INDEX_LSH);
                this->matcher = cv::makePtr<cv::FlannBasedMatcher>(indexParams, searchParams);
                */
            }
        }

        if(this->detector != NULL && this->extractor != NULL) {
            if(this->detector->descriptorType() != this->extractor->descriptorType()) {
                this->detector = cv::Ptr<cv::FeatureDetector>(); // NULL
                this->extractor = cv::Ptr<cv::DescriptorExtractor>(); // NULL
            }
        }


#ifdef USE_GPU
        gpu_surf = new cv::gpu::SURF_GPU(minHessian);
#endif
    }

    cv::Mat Besra::readImage(const fs::path &file) {
        return cv::imread(file.string(), cv::IMREAD_GRAYSCALE);
    }

    std::vector<cv::KeyPoint> Besra::detectKeypoints(const cv::Mat &img) {
        std::vector<cv::KeyPoint> keypoints; 

#ifdef USE_GPU
        cv::gpu::GpuMat gpu_keypoints;
        cv::gpu::GpuMat gpu_descriptors;
        cv::gpu::GpuMat gpu_img;
        gpu_img.upload(img);
        (*gpu_surf)(gpu_img, cv::gpu::GpuMat(), gpu_keypoints, gpu_descriptors);
        gpu_surf->downloadKeypoints(gpu_keypoints, keypoints); 
#else
        detector->detect(img, keypoints);
#endif

        return keypoints;
    }

    cv::Mat Besra::detectAndCompute(const cv::Mat &img) {
        cv::Mat descriptors;

#ifdef USE_GPU
        cv::gpu::GpuMat gpu_keypoints;
        cv::gpu::GpuMat gpu_descriptors;
        cv::gpu::GpuMat gpu_img;
        gpu_img.upload(img);
        (*gpu_surf)(gpu_img, cv::gpu::GpuMat(), gpu_keypoints, gpu_descriptors);
        gpu_descriptors.download(descriptors);
#else
        std::vector<cv::KeyPoint> keypoints = detectKeypoints(img);
        extractor->compute(img, keypoints, descriptors);
#endif

        return descriptors;
    }

    cv::Mat Besra::detectAndCompute(const cv::Mat &img, cv::Ptr<cv::BOWImgDescriptorExtractor> bow) {
        cv::Mat descriptors;
            
        std::vector<cv::KeyPoint> keypoints = detectKeypoints(img);
        bow->compute(img, keypoints, descriptors);

        return descriptors;
    }

    cv::Mat Besra::buildVocabulary(const fs::path &input_file, int clusterCount, int threads) {
        cv::BOWKMeansTrainer bowtrainer(clusterCount);

        std::pair<cv::Mat, cv::Mat> result = processImages(input_file, threads);
        cv::Mat descriptors = result.first;

        int type = descriptors.type();
        if(type != CV_32F) {
            descriptors.convertTo(descriptors, CV_32F);
        }

        for(int i = 0; i < descriptors.rows; i++) {
            bowtrainer.add(descriptors.row(i));
        }

        if(bowtrainer.descriptorsCount() == 0) {
            //TODO: throw exception or bail here
            BOOST_LOG_TRIVIAL(error) << "<buildVocabulary> No descriptors found! Can't perform clustering..";
        }
        
        BOOST_LOG_TRIVIAL(info) << "<buildVocabulary> Clustering.." << bowtrainer.getDescriptors().size();
        cv::Mat vocabulary = bowtrainer.cluster();

        if(type != CV_32F) {
            vocabulary.convertTo(vocabulary, type);
        }

        return vocabulary;
    }

    bool Besra::processLine(std::string line, cv::Mat &descriptors, int &label,
                            cv::Ptr<cv::BOWImgDescriptorExtractor> bow) {
        std::vector<std::string> tokens;  
        boost::split(tokens, line, boost::is_any_of("\t,"));
        if(tokens.size() != 2) {
            BOOST_LOG_TRIVIAL(error) <<  "Invalid record format for line: " << line;
            return false;
        }

        fs::path img_path(tokens[0]);
        if(!fs::is_regular_file(img_path)) return false;

        label = 0;
        try {
            label = std::stoi(tokens[1]);
        } catch ( ... ) {
            BOOST_LOG_TRIVIAL(error) <<  "Invalid label for image file: " << img_path;
            return false;
        }

        cv::Mat img = readImage(img_path);
        if(bow == NULL) { 
            descriptors = detectAndCompute(img);
        } else {
            descriptors = detectAndCompute(img, bow);
        }

        if(descriptors.empty()) {
            BOOST_LOG_TRIVIAL(warning) <<  "Empty descriptors for image: " << img_path;
            return false;
        }

        return true;
    }

    std::pair<cv::Mat, cv::Mat> Besra::processImages(const fs::path &path, int threads, 
                                                     cv::Ptr<cv::BOWImgDescriptorExtractor> bow) {
        cv::Mat descriptors;
        cv::Mat labels;

        if(threads <= 0) {
            threads = 1;
        }

        int count = 0;
        std::ifstream ifs(path.c_str());
        #pragma omp parallel num_threads(threads)
        {
            int tcount = 0;
            std::string line;
            while(true) {

                #pragma omp critical(input)
                {
                    std::getline(ifs, line);
                }
                if(ifs.eof()) break;


                cv::Mat d;
                int label;
                if(processLine(line, d, label, bow)) {
                    tcount++;
                    #pragma omp critical
                    {
                        descriptors.push_back(d);
                        labels.push_back(label);
                        count++;
                    }
                }

                if(tcount % 100 == 0) {
#ifdef OPENMP_FOUND
                    BOOST_LOG_TRIVIAL(info) <<  "Thread " << omp_get_thread_num() << " progress: " << tcount;
#else
                    BOOST_LOG_TRIVIAL(info) <<  "progress: " << tcount;
#endif
                }
            }
        }

        return std::make_pair(descriptors, labels);
    }

    cv::Ptr<cv::BOWImgDescriptorExtractor> Besra::loadBOW(const cv::Mat &vocabulary) {
        cv::Ptr<cv::BOWImgDescriptorExtractor> bow = cv::makePtr<cv::BOWImgDescriptorExtractor>(extractor, matcher);
        bow->setVocabulary(vocabulary);
        return bow;
    }

    cv::Ptr<cv::ml::StatModel> Besra::train(const fs::path &input_file, const cv::Mat &vocabulary, int threads) {
        cv::Ptr<cv::BOWImgDescriptorExtractor> bow = loadBOW(vocabulary);

        cv::Mat samples;
        cv::Mat labels;

        std::pair<cv::Mat, cv::Mat> result = processImages(input_file, threads, bow);
        samples = result.first;
        if(samples.type() != CV_32F) {
            samples.convertTo(samples, CV_32F);
        }

        labels = result.second;

        BOOST_LOG_TRIVIAL(info) << "Training StatsModel";

        if(this->bayes) {
            cv::Ptr<cv::ml::TrainData> trainData = cv::ml::TrainData::create(samples, cv::ml::ROW_SAMPLE, labels);
            return cv::ml::StatModel::train<cv::ml::NormalBayesClassifier>(trainData);
        }

        cv::Ptr<cv::ml::SVM> svm = cv::ml::SVM::create();
        svm->setType(cv::ml::SVM::C_SVC);
        svm->setKernel(cv::ml::SVM::LINEAR);
        svm->setTermCriteria(cv::TermCriteria(cv::TermCriteria::MAX_ITER+cv::TermCriteria::EPS, 1000, FLT_EPSILON));

        svm->train(samples, cv::ml::ROW_SAMPLE, labels);
        cv::Ptr<cv::ml::StatModel> mod = svm;
        return mod;
    }

    float Besra::classify(const fs::path &path, cv::Ptr<cv::BOWImgDescriptorExtractor> bow, cv::Ptr<cv::ml::StatModel> model) {
        cv::Mat img = readImage(path);
        cv::Mat features = detectAndCompute(img, bow);

        if(features.empty()) {
            throw std::out_of_range("Empty features");
        }

        float res = model->predict(features);
        return res;
    }

    cv::Ptr<cv::ml::StatModel> Besra::loadStatModel(const fs::path &cache, const cv::Mat &vocabulary) {
        if(this->bayes) {
            return cv::ml::StatModel::load<cv::ml::NormalBayesClassifier>(cache.string());
        }
        return cv::ml::StatModel::load<cv::ml::SVM>(cache.string());
    }

}


src/besra.hpp
/**
 * Copyright (C) 2014 Andrew E. Bruno
 *
 * This file is part of Besra
 *
 * Besra is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#ifndef __BESRA_H_INCLUDED__
#define __BESRA_H_INCLUDED__ 


#include <iostream>
#include <fstream>
#include <utility>
#include <string>
#include <queue>
#include <stdexcept>
#include <boost/algorithm/string.hpp>
#include <boost/log/trivial.hpp>
#include <boost/log/utility/setup.hpp>
#include <boost/filesystem.hpp>
#include <opencv2/opencv.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/ml/ml.hpp>
#include <opencv2/features2d/features2d.hpp>
#include <opencv2/xfeatures2d/nonfree.hpp>
#include <opencv2/xfeatures2d.hpp>
#include <opencv2/highgui/highgui.hpp>
#ifdef USE_GPU
#include <opencv2/gpu/gpu.hpp>
#endif
#ifdef OPENMP_FOUND
#include <omp.h>
#endif

namespace fs = boost::filesystem;

namespace besra {

    void init_log(bool verbose=false);

    class Besra {
        private:
            bool processLine(std::string line, cv::Mat &descriptors, int &label,
                              cv::Ptr<cv::BOWImgDescriptorExtractor> bow);

            std::pair<cv::Mat, cv::Mat> processImages(const fs::path &file, int threads = 0, 
                                                      cv::Ptr<cv::BOWImgDescriptorExtractor> bow = cv::Ptr<cv::BOWImgDescriptorExtractor>());

        public:
            bool bayes;
            cv::Ptr<cv::DescriptorMatcher> matcher;
            cv::Ptr<cv::DescriptorExtractor> extractor;
            cv::Ptr<cv::FeatureDetector> detector;
#ifdef USE_GPU
            cv::Ptr<cv::gpu::SURF_GPU> gpu_surf;
#endif

            Besra(int minHessian = 600, std::string extractor="SURF", std::string detector="SURF", bool bayes = false);

            cv::Mat readImage(const fs::path &file);

            std::vector<cv::KeyPoint> detectKeypoints(const cv::Mat &img);
            cv::Mat detectAndCompute(const cv::Mat &img);
            cv::Mat detectAndCompute(const cv::Mat &img, cv::Ptr<cv::BOWImgDescriptorExtractor> bow);

            cv::Mat buildVocabulary(const fs::path &input_file, int clusterCount = 150, int threads = 0);
            cv::Ptr<cv::BOWImgDescriptorExtractor> loadBOW(const cv::Mat &vocabulary);
            cv::Ptr<cv::ml::StatModel> train(const fs::path &input_file, const cv::Mat &vocabulary, int threads = 0);
            cv::Ptr<cv::ml::StatModel> loadStatModel(const fs::path &cache, const cv::Mat &vocabulary);
            float classify(const fs::path &path, cv::Ptr<cv::BOWImgDescriptorExtractor> bow, cv::Ptr<cv::ml::StatModel> model);
    };

}

#endif


src/besra-view.cpp
/**
 * Copyright (C) 2014 Andrew E. Bruno
 *
 * This file is part of Besra
 *
 * Besra is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "besra.hpp"
#include <boost/program_options.hpp>
#include <boost/tokenizer.hpp>

namespace po = boost::program_options;

int main(int argc, char** argv) {
    besra::init_log();

    std::string img_path;
    std::string detector_str;
    int minHessian;
    po::options_description desc("Options");
    desc.add_options()
        ("help,h", "help message")
        ("hessian,k", po::value<int>(&minHessian)->default_value(600), "hessian threshold")
        ("image,i", po::value<std::string>(&img_path)->required(), "path to image file")
        ("detector,d", po::value<std::string>(&detector_str)->default_value("SURF"), "feature detector (SURF, BRISK, ORB, KAZE, AKAZE, MSER, SIFT, FAST, GFTT, BLOB)")
    ;

    po::variables_map vm;
    try {
        po::store(po::parse_command_line(argc, argv, desc), vm);

        if (vm.count("help")) {
            std::cout << desc << "\n";
            return 1;
        }

        po::notify(vm);    
    } catch(po::error& e) { 
      std::cerr << "ERROR: " << e.what() << std::endl << std::endl; 
      std::cerr << desc << std::endl; 
      return 1; 
    } 

    fs::path img_file(img_path);

    if(!fs::exists(img_file)) {
      std::cerr << "Invalid image file: " << img_file << std::endl; 
      std::cerr << desc << std::endl; 
      return 1; 
    }

#ifdef USE_GPU
    int gpus = cv::gpu::getCudaEnabledDeviceCount();
    if(gpus > 0) {
        BOOST_LOG_TRIVIAL(info) << "Found " <<  gpus << " GPUs";
    } else {
        std::cerr << "USE_GPU is enabled and no GPUs found!" << std::endl; 
        return 1; 
    }
#endif

    besra::Besra besra(minHessian, "SURF", detector_str); 
    if(besra.detector == NULL) {
        std::cerr << "Invalid detector: " << detector_str << std::endl; 
        return 1; 
    }

    cv::Mat img = besra.readImage(img_file);
    std::vector<cv::KeyPoint> keypoints = besra.detectKeypoints(img);

    cv::Mat img_keypoints;
      
    cv::drawKeypoints(img, keypoints, img_keypoints, cv::Scalar::all(-1), cv::DrawMatchesFlags::DEFAULT);
    cv::imshow("Keypoints: " + img_file.filename().string(), img_keypoints);

    cv::waitKey(0);

    return 0;
}


src/besra-trainer.cpp
/**
 * Copyright (C) 2014 Andrew E. Bruno
 *
 * This file is part of Besra
 *
 * Besra is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "besra.hpp"
#include "boost/program_options.hpp"

namespace po = boost::program_options;

int main(int argc, char** argv) {

    fs::path cwd = fs::initial_path();
    int clusters;
    int threads;
    int minHessian;
    bool verbose;
    std::string classifier;
    std::string input_path;
    std::string output_path;
    std::string detector_str;
    std::string extractor_str;

    po::options_description desc("Options");
    desc.add_options()
        ("help,h", "help message")
        ("input,i", po::value<std::string>(&input_path)->required(), "path to input file")
        ("output,o", po::value<std::string>(), "path to output directory")
        ("threads,t", po::value<int>(&threads)->default_value(1), "number of threads to spawn")
        ("clusters,c", po::value<int>(&clusters)->default_value(150), "clusters")
        ("hessian,k", po::value<int>(&minHessian)->default_value(600), "hessian threshold")
        ("vocab,b", po::value<std::string>(), "path to vocabulary cache file")
        ("verbose,v", po::bool_switch(&verbose)->default_value(false), "verbose output")
        ("classifier,j", po::value<std::string>(&classifier)->default_value("svm"), "classifier (svn, bayes)")
        ("detector,d", po::value<std::string>(&detector_str)->default_value("SURF"), "feature detector (SURF, BRISK, ORB, KAZE, AKAZE, MSER, SIFT, FAST, GFTT, BLOB)")
        ("extractor,e", po::value<std::string>(&extractor_str)->default_value("SURF"), "descriptor extractor (SURF, FREAK, BRISK, ORB, KAZE, AKAZE, BRIEF, SIFT)")
    ;

    po::variables_map vm;
    try {
        po::store(po::parse_command_line(argc, argv, desc), vm);

        if (vm.count("help")) {
            std::cout << desc << "\n";
            return 1;
        }

        po::notify(vm);    
    } catch(po::error& e) { 
      std::cerr << "ERROR: " << e.what() << std::endl << std::endl; 
      std::cerr << desc << std::endl; 
      return 1; 
    } 

    besra::init_log(verbose);

    fs::path input_file(input_path);
    fs::path output_dir(cwd);

    if(!fs::is_regular_file(input_file)) {
      std::cerr << "Invalid input file: " << input_file << std::endl; 
      return 1; 
    }

    if(vm.count("output")) {
        output_dir = fs::path(vm["output"].as<std::string>());
        if( !fs::exists(output_dir) || !fs::is_directory(output_dir)) {
            std::cerr << "Invalid output directory: " << output_dir << std::endl; 
            return 1; 
        }
    }

    fs::path model_cache_file(output_dir / fs::path("stats-model.yml"));
    fs::path vocab_cache_file(output_dir / fs::path("bow-vocab.yml"));
    if(vm.count("vocab")) {
        vocab_cache_file = fs::path(vm["vocab"].as<std::string>());
        if( !fs::exists(vocab_cache_file) ) {
            std::cerr << "Invalid vocabulary cache file: " << vocab_cache_file << std::endl; 
            return 1; 
        }
    }


#ifdef USE_GPU
    int gpus = cv::gpu::getCudaEnabledDeviceCount();
    if(gpus > 0) {
        BOOST_LOG_TRIVIAL(info) << "Found " <<  gpus << " GPUs";
    } else {
        std::cerr << "USE_GPU is enabled and no GPUs found!" << std::endl; 
        return 1; 
    }
#endif
    besra::Besra besra(minHessian, extractor_str, detector_str); 
    if(besra.extractor == NULL || besra.detector == NULL) {
        std::cerr << "Invalid detector/extractor combination: " << detector_str << " / " << extractor_str << std::endl; 
        return 1; 
    }

    if(extractor_str != "SURF" || detector_str != "SURF") {
        // Only SURF appears to be thread safe at the moment
        threads = 1;
    }

    if(classifier == "bayes") {
        besra.bayes = true;
    }

    cv::Mat vocabulary;

    if(vm.count("vocab")) {
        BOOST_LOG_TRIVIAL(info) << "Loading vocab from file: " << vocab_cache_file.string();
        cv::FileStorage fs(vocab_cache_file.string(), cv::FileStorage::READ);
        fs["vocabulary"] >> vocabulary;
        fs.release();   
    } else {
        BOOST_LOG_TRIVIAL(info) << "Building vocab..";
        vocabulary = besra.buildVocabulary(input_file, clusters, threads);

        BOOST_LOG_TRIVIAL(info) << "Saving vocab to cache file: " << vocab_cache_file.string();
        cv::FileStorage vocab_fs(vocab_cache_file.string(), cv::FileStorage::WRITE);
        vocab_fs << "vocabulary" << vocabulary;
        vocab_fs.release();
    }


    BOOST_LOG_TRIVIAL(info) << "Building stats model..";
    cv::Ptr<cv::ml::StatModel> model = besra.train(input_file, vocabulary, threads);

    BOOST_LOG_TRIVIAL(info) << "Saving stats model to cache file: " << model_cache_file.string();
    model->save(model_cache_file.string());

    BOOST_LOG_TRIVIAL(info) << "Done!";

    return 0;
}
